src/NHibernate/Collection/AbstractPersistentCollection.cs

```diff
@@ -716,29 +717,37 @@ protected virtual ICollection GetOrphans(ICollection oldElements, ICollection cu
 				return oldElements;
 			}
 
-			IType idType = session.Factory.GetEntityPersister(entityName).IdentifierType;
+			if (currentElements.Cast<object>().SequenceEqual(oldElements.Cast<object>()))
+				return Array.Empty<object>();
 
-			// create the collection holding the orphans
-			List<object> res = new List<object>();
+			var persister = session.Factory.GetEntityPersister(entityName);
+			IType idType = persister.IdentifierType;
 
-			// collect EntityIdentifier(s) of the *current* elements - add them into a HashSet for fast access
-			var currentIds = new HashSet<TypedValue>();
+			var currentObjects = new HashSet<object>(idType.GetComparer());
 			foreach (object current in currentElements)
 			{
-				if (current != null && ForeignKeys.IsNotTransientSlow(entityName, current, session))
+				if (current != null)
 				{
-					object currentId = ForeignKeys.GetEntityIdentifierIfNotUnsaved(entityName, current, session);
```

Comment:
This call checks session entry dictionary and throw exception for transient object. This exception doesn't make any sense  for orphans detection - we should simply ignore transient objects. So let's call `persister.GetIdentifier` directly without any additional checks